upgrades/steps_28.go

```diff
@@ -139,59 +156,163 @@ var getUpgradeStepsClient = func(caller base.APICaller) UpgradeStepsClient {
 		// e.g. /var/lib/juju/agents/unit-ubuntu-0/state/uniter
-		uniterStateFile := filepath.Join(agent.BaseDir(dataDir), unitName, "state", "uniter")
+		uniterStateDir := filepath.Join(agent.BaseDir(dataDir), unitName, "state")
 
-		// Read into a State as we can validate the data.
-		var st operation.State
-		if err = utils.ReadYaml(uniterStateFile, &st); err != nil {
-			// No file available, move on.
-			if os.IsNotExist(err) {
-				continue
-			}
-			return nil, errors.Annotatef(err, "failed to read uniter state from %q", uniterStateFile)
+		uniterSt, filename, err := readUniterState(uniterStateDir)
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
 		}
-		fileNames = append(fileNames, uniterStateFile)
-		if err = st.Validate(); err != nil {
-			return nil, errors.Annotatef(err, "failed to validate uniter state from %q", uniterStateFile)
+		if uniterSt != nil {
+			fileNames.Add(filename)
+			uniterStates[i].UniterState = uniterSt
 		}
 
-		data, err := yaml.Marshal(st)
-		if err != nil {
-			return nil, errors.Annotatef(err, "failed to unmarshal uniter state from %q", uniterStateFile)
+		storageSt, filename, err := readStorageState(uniterStateDir)
+		if err != nil && !os.IsNotExist(err) {
+			return nil, err
+		}
+		if storageSt != nil {
+			fileNames.Add(filename)
+			uniterStates[i].StorageState = storageSt
 		}
-		uniterStates[tag] = string(data)
 	}
 
+	logger.Criticalf("filenames %d %s", fileNames.Size(), fileNames.Values())
 	// No state files, nothing to do.
-	if len(fileNames) == 0 {
+	if fileNames.IsEmpty() {
 		return nil, nil
 	}
 
 	client := getUpgradeStepsClient(ctx.APIState())
 	if err := client.WriteUniterState(uniterStates); err != nil {
 		return nil, errors.Annotatef(err, "unable to set state for units %q", strings.Join(unitNames, ", "))
 	}
-	return fileNames, nil
+	return fileNames.Values(), nil
+}
+
+func readUniterState(uniterStateDir string) (*string, string, error) {
+	uniterStateFile := filepath.Join(uniterStateDir, "uniter")
+	// Read into a State as we can validate the data.
+	var st operation.State
+	if err := utils.ReadYaml(uniterStateFile, &st); err != nil {
+		// No file available, move on.
+		if os.IsNotExist(err) {
+			return nil, "", err
+		}
+		return nil, "", errors.Annotatef(err, "failed to read uniter state from %q", uniterStateFile)
+	}
+	if err := st.Validate(); err != nil {
+		return nil, "", errors.Annotatef(err, "failed to validate uniter state from %q", uniterStateFile)
+	}
+
+	data, err := yaml.Marshal(st)
+	if err != nil {
+		return nil, "", errors.Annotatef(err, "failed to unmarshal uniter state from %q", uniterStateFile)
+	}
+	str := string(data)
+
+	return &str, uniterStateFile, nil
 }
 
 func caasOperatorLocalState(ctx Context, unitNames []string) ([]string, error) {
 	// TODO: (hml) 27-Mar-2020.
 	// Implement when relations, storage, metrics and the operator state are moved.
 	return nil, nil
 }
+
+func readStorageState(uniterStateDir string) (*string, string, error) {
```

Comment:
Same comment as above about returning pointers to scalars.

Reply:
logic changed